src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -1,138 +1,433 @@
+
+      // isAllChecked defines the isChecked status of parent checkboxes. If all child items are selected, it will turn true, otherwise it returns false.
+      // This prop is only passed to parent options.
+      const isAllChecked = (
+        multiSelectId: string,
+        item: MultiSelectItem
+      ): boolean => {
+        let childIds: string[] = item.children.map((childItem) => childItem.id);
+        if (selectedItems[multiSelectId] !== undefined) {
+          return childIds.every(
+            (childItem) =>
+              selectedItems[multiSelectId].items.includes(childItem) &&
+              !item.children.find((c) => c.id === childItem)?.isDisabled
+          );
+        }
+        return false;
+      };
+
+      // isInteterminate will return true if some child items of the parent item are selected. This prop is only passed to parent options.
+      const isIndeterminate = (
+        multiSelectId: string,
+        item: MultiSelectItem
+      ): boolean => {
+        let childIds: string[] = item.children.map((childItem) => childItem.id);
+        if (
+          selectedItems[multiSelectId] !== undefined &&
+          childIds.length > 0 &&
+          childIds.some(
+            (childItem) =>
+              selectedItems[multiSelectId].items.includes(childItem) &&
+              !item.children.find((c) => c.id === childItem)?.isDisabled
+          )
+        ) {
+          return !isAllChecked(multiSelectId, item);
+        }
+        return false;
+      };
+
+      const updateDisabledState = (
+        multiSelectId: string,
+        item: MultiSelectItem
+      ): boolean => {
+        // Update isDisabled for the parent item
+        item.isDisabled = isAllChecked(multiSelectId, item);
+
+        // Check if all child items are disabled, and update the parent accordingly
+        if (item.children.every((childItem) => childItem.isDisabled)) {
+          item.isDisabled = true;
+        } else {
+          item.isDisabled = false;
+        }
+        // Return the updated isDisabled state for the parent item
+        return item.isDisabled;
       };
 
-      if (type === "listbox") {
-        const listboxOnChange = onChange as ListboxOnChange;
+      const onChangeSearch = (event) => {
+        const value = event.target.value.trim().toLowerCase();
 
+        if (value === "") {
+          return listItemsCount === defaultItemsVisible
+            ? displayDefaultItems()
+            : setItemsList(items);
+        }
+
+        const filteredItems = items.filter((item) => {
+          if (item.children) {
+            if (item.name.toLowerCase().includes(value)) {
+              return true;
+            }
+
+            return item.children.some((childItem) =>
+              childItem.name.toLowerCase().includes(value)
+            );
+          }
+
+          return item.name.toLowerCase().includes(value);
+        });
+
+        setItemsList(filteredItems);
+      };
+
+      const clearSearchKeyword = () => {
```

Comment:
Can you add a description to this function? For example, when do you need to reset the list of items when the search is cleared?